src/storage/v2/indices.cpp

```diff
@@ -263,6 +265,95 @@ bool CurrentVersionHasLabelProperty(const Vertex &vertex, LabelId label, Propert
   return !deleted && has_label && current_value_equal_to_value;
 }
 
+template <typename TIndexAccessor>
+void TryInsertLabelIndex(Vertex &vertex, LabelId label, TIndexAccessor &index_accessor) {
+  if (vertex.deleted || !utils::Contains(vertex.labels, label)) {
+    return;
+  }
+
+  index_accessor.insert({&vertex, 0});
+}
+
+template <typename TIndexAccessor>
+void TryInsertLabelPropertyIndex(Vertex &vertex, std::pair<LabelId, PropertyId> label_property_pair,
+                                 TIndexAccessor &index_accessor) {
+  if (vertex.deleted || !utils::Contains(vertex.labels, label_property_pair.first)) {
+    return;
+  }
+  auto value = vertex.properties.GetProperty(label_property_pair.second);
+  if (value.IsNull()) {
+    return;
+  }
+  index_accessor.insert({std::move(value), &vertex, 0});
+}
+
+template <typename TSkiplistIter, typename TIndex, typename TIndexKey, typename TFunc>
+void CreateIndexOnSingleThread(utils::SkipList<Vertex>::Accessor &vertices, TSkiplistIter it, TIndex &index,
+                               TIndexKey key, const TFunc &func) {
+  utils::MemoryTracker::OutOfMemoryExceptionEnabler oom_exception;
+  try {
+    auto acc = it->second.access();
+    for (Vertex &vertex : vertices) {
+      func(vertex, key, acc);
+    }
+  } catch (const utils::OutOfMemoryException &) {
+    utils::MemoryTracker::OutOfMemoryExceptionBlocker oom_exception_blocker;
+    index.erase(it);
+    throw;
+  }
+}
+
+template <typename TIndex, typename TIndexKey, typename TSKiplistIter, typename TFunc>
+void CreateIndexOnMultipleThreads(utils::SkipList<Vertex>::Accessor &vertices, TSKiplistIter skiplist_iter,
+                                  TIndex &index, TIndexKey key, const ParalellizedIndexCreationInfo &paralell_exec_info,
+                                  const TFunc &func) {
+  utils::MemoryTracker::OutOfMemoryExceptionEnabler oom_exception;
+
+  const auto &vertex_batches = paralell_exec_info.first;
+  const auto thread_count = paralell_exec_info.second;
+
+  MG_ASSERT(!vertex_batches.empty(),
+            "The size of batches should always be greater than zero if you want to use the parallel version of index "
+            "creation!");
+
+  std::atomic<uint64_t> batch_counter = 0;
+
+  utils::Synchronized<std::optional<utils::OutOfMemoryException>, utils::SpinLock> maybe_error{};
+  {
+    std::vector<std::jthread> threads;
+    threads.reserve(thread_count);
+
+    for (auto i{0U}; i < thread_count; ++i) {
+      threads.emplace_back(
+          [&skiplist_iter, &func, &index, &vertex_batches, &maybe_error, &batch_counter, &key, &vertices]() {
+            while (!maybe_error.Lock()->has_value()) {
+              const auto batch_index = batch_counter++;
+              if (batch_index >= vertex_batches.size()) {
+                return;
```

Comment:
Instead of creating more threads then are batches, you can set thread count to `min(thread_count, batch_size)`. This way you don't need to return here.

Reply:
That is a good point, however that `return` statement is there because of the concurrency of the threads. The batch_index is used to calculate which batch the given thread will work on next. If you have 10 worker threads and 15 batches overall, lets say the 10 threads will take the first 10 batches in some order. Individual threads will finish the processing of their assigned batches after some arbitrary time. We do not have a way to know if the there are still work left to do (batches left to process), so when a given thread asks for another batch to process(gets the value of the atomic counter), we have to make sure that the batch represented by the integer value is still actually a batch that is needed to be processed. If our thread wakes up and batch_index it gets is 15 in this specific case we know that this integer is not representing another batch, since `batch_index >= vertex_batches.size()`.

There is an argument to be made in cases where `thread_count  > batches.size()`, however since the parallel index creation is a settable flag, which will be false by default, I would say lets give that heuristic decision to the user. This should be only used in cases where we try to recover a huge set of data, where chances of `thread_count  > batches.size()` are slim.

Let me know if you disagree or if I misunderstood something.

Reply:
Oh I was wrong, you still need return here :smile:

Reply:
I was thinking something else.

On these two lines, it is possible that vertex_batches.size() > thread_count, so we reserve and create more threads than needed.

```
std::vector<std::jthread> threads;
threads.reserve(thread_count);
```
We need a return statement, but I was thinking that we can set `thread_count = min(thread_count, vertex_batches.size())`. What do you think about that?

Reply:
Yep, I will do that.